### ISSUE TITLE
Cache data race / buffer overwrite fix

### ISSUE THREAD

**Issue by majodev** [darmiel/yaxc#23]
Hi,

in response to your [StackOverflow question](https://stackoverflow.com/questions/66930097/race-with-mutex-corrupt-data-in-map) and [Data Race in Cache #8](https://github.com/darmiel/yaxc/issues/8), see [majodev/yaxc ](https://github.com/majodev/yaxc) for a minimal reproduction test and fix.

TL;DR:
[Go fiber zero allocation](https://docs.gofiber.io/#zero-allocation). The byte slice buffer that holds your strings `path` and `hash` which you got from `ctx` is only valid within the handler. The key/values stored inside your cache map *will change* as these `strings` are still based on this buffer that gets reused in each handler. 

Either: 
* Create a new Buffer from the read values inside the handler and copy its contents over 
* or use the `utils.ImmutableString` util (same as above)
* or use the `Immutable` config flag (perf penalty, no longer zero allocated context).

**Comment by darmiel** [comment]
Hello @majodev,

first of all: **thank you very much for your help and your work! 💙**
I didn't expect anyone to look seriously at the code because of the problem.

Your fork also gives me another look at how to write tests for a web app, and how to reproduce bugs more easily. 
So, thanks again for that. I saved the fork (/source code), which can be useful for me again in the future :).

But about the bug:
I was not aware that values of gofiber are not immutable by default. Even though the hint is right on the home page, I guess I didn't really notice it or understand the problem.

I have now implemented it as in your suggested solution (0c764a737e2bbeea8f417e05af6214589dc66e78).
However, I have changed one thing: Instead of `utils.ImmutableString()` I used `utils.CopyString()` because `utils.ImmutableString()` is shown as deprecated according to the [documentation](https://pkg.go.dev/github.com/gofiber/fiber/v2/utils@v2.6.0#ImmutableString).

---

**About the other solutions:**
Do you mean something like this?

> Create a new Buffer from the read values inside the handler and copy its contents over
```go
buf := bytes.NewBufferString(ctx.Params("hash"))
hash := string(buf.Bytes())
```

> or use the Immutable config flag
```go
cfg := &fiber.Config{Immutable: true}
```

**Comment by majodev** [comment]
Hi,

you are welcome. Especially thank you for your [StackOverflow Question](https://stackoverflow.com/questions/66930097/race-with-mutex-corrupt-data-in-map) - we are currently also in the progress to track down a similar (but sightly more complex) issue in a customer project and your question helped to point us to the right direction. 😇
